### PR TITLE
mkl: fix .pc file, add test

### PR DESCRIPTION
**Pull request by danieldk**
###### Motivation for this change

There was a spurious `intel64` in some library paths, which is fixed
in the first commit.

The MKL pkg-config files often change and are then incorrect for our
paths. pkg-config validation finds some issues, but not incorrect
paths. The second commit adds a small test program to test whether the
generated pkg-config files can actually be used to build a functioning
binary. Hopefully this catches future regressions.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/science/math/mkl/default.nix [hunk 1]
```diff
@@ -156,6 +158,8 @@ in stdenvNoCC.mkDerivation {
   dontStrip = true;
   dontPatchELF = true;
 
+  passthru.tests = callPackage ./test { };
```
Comment (jtojnar) [inline pkgs/development/libraries/science/math/mkl/default.nix:161]: This should ideally be attribute set containing tests.

Comment (danieldk) [reply pkgs/development/libraries/science/math/mkl/default.nix]: Ah, right, thanks! Fixed.

File: pkgs/development/libraries/science/math/mkl/test/default.nix [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+  buildPhase = ''
+    # Check regular Nix build.
+    gcc $(pkg-config --cflags --libs mkl-dynamic-ilp64-seq) test.c -o test
+
+    # Clear flags to ensure that we are purely relying on options
+    # provided by pkg-config.
+    NIX_CFLAGS_COMPILE="" \
+    NIX_LDFLAGS="" \
+      gcc $(pkg-config --cflags --libs mkl-dynamic-ilp64-seq) test.c -o test
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/science/math/mkl/test/default.nix:23]:
```suggestion
    NIX_LDFLAGS="" \
    gcc $(pkg-config --cflags --libs mkl-dynamic-ilp64-seq) test.c -o test
```

Comment (danieldk) [reply pkgs/development/libraries/science/math/mkl/test/default.nix:23]: I don't know, I like the original formatting better, it sets apart the variables from the command.

Comment (SuperSandro2000) [reply pkgs/development/libraries/science/math/mkl/test/default.nix:23]: It is inconsistent with the line above. Either intend the last 2 lines or intent nothing.

**Comment by danieldk** [comment]
@bhipple does this look ok to you?

**Review by bhipple — APPROVED** [review]

**Comment by bhipple** [comment]
Looks nice, ty as always for increasing the amount of automation we have here.  

I wonder if the generic pkg-config checker could be enhances to run something like `pkg-config --cflags --libs *.pc` and verify all the paths on `-L/foo -I/bar` actually exist ...

**Comment by danieldk** [comment]
> I wonder if the generic pkg-config checker could be enhances to run something like `pkg-config --cflags --libs *.pc` and verify all the paths on `-L/foo -I/bar` actually exist ...

I think that would be a really nice feature to have. `pkg-config` validation is really limited. I recently looked a bit at parsing `pc` files directly, but I didn't have much time to investigate how tricky variable substitution would be. But just calling `pkg-config` and parsing the library paths might be a good alternative.

**Pull request merged**